### PR TITLE
fix(context): join all system messages into instructions (#390)

### LINKED ISSUE #390: Every system message after the first is silently discarded
## Problem

Every `system` message after the first is silently discarded before token counting and generation.

`ContextManager.makeSession` strips **all** system messages out of the conversation, and
`buildInstructions` then copies only the **first** one back into the transcript. Everything in
between is gone, with no warning, no error, and no token accounting.

## Where

`Sources/ContextManager.swift:35`, `Sources/ContextManager.swift:127-130` (commit b39659abe6b627e95454c8144d8fdeff9b3f50a3)

The same path serves Responses `developer` messages (`Sources/Core/ResponsesModels.swift:140-158`)
and CLI `--messages` conversations combined with `-s`.

## Evidence

Token accounting proves the message never reaches the model. Two requests to `/v1/chat/completions`
differing only by an extra ~400-word system message:

```
[system "Be concise.",                          user "Say OK."]  -> prompt_tokens 28
[system "Be concise.", system <400 words>,      user "Say OK."]  -> prompt_tokens 28
```

A 400-word instruction contributes exactly zero tokens.

## Impact

Multi-system-message conversations are ordinary: SDKs and frameworks routinely emit a base system
prompt plus a task-specific one, and LangChain/OpenAI-style agents stack them. apfel accepts the
request, returns HTTP 200, and answers while ignoring instructions the caller believes are in
force. This is silent, not loud - the caller has no way to detect it short of counting tokens.

## Reproduction

With `apfel --serve --port 11434` running:

```sh
python3 - <<'PY'
import json, urllib.request
def post(msgs):
    req = urllib.request.Request('http://127.0.0.1:11434/v1/chat/completions',
        data=json.dumps({'model':'apple-foundationmodel','messages':msgs,'max_tokens':8}).encode(),
        headers={'Content-Type':'application/json'})
    with urllib.request.urlopen(req, timeout=60) as r:
        return json.load(r)['usage']['prompt_tokens']
base  = [{'role':'system','content':'Be concise.'}, {'role':'user','content':'Say OK.'}]
extra = [{'role':'system','content':'Be concise.'},
         {'role':'system','content':'Additional instruction. ' * 100},
         {'role':'user','content':'Say OK.'}]
print('one system :', post(base))
print('two systems:', post(extra))
PY
```

Observed: `one system : 28` and `two systems: 28`.

## Concrete fix

Join every system message into the instructions block instead of taking only the first.

```diff
--- a/Sources/ContextManager.swift
+++ b/Sources/ContextManager.swift
@@ -125,8 +125,15 @@
         // System prompt
-        if let sys = messages.first(where: { $0.role == "system" })?.textContent {
-            parts.append(sys)
+        // Every system message must survive, not just the first: makeSession
+        // filters them all out of the conversation (line 35), so any system
+        // message not copied here is silently dropped before token counting
+        // and generation.
+        let systemParts = messages
+            .filter { $0.role == "system" }
+            .compactMap { $0.textContent }
+            .filter { !$0.isEmpty }
+        if !systemParts.isEmpty {
+            parts.append(systemParts.joined(separator: "\n\n"))
         }
```

## Acceptance criteria

- [ ] Two system messages produce a strictly larger `prompt_tokens` than one
- [ ] The content of every system message is present in the built instructions, in order
- [ ] A conversation with zero system messages behaves exactly as today
- [ ] Responses `developer` messages and CLI `--messages` + `-s` get the same treatment
- [ ] If a future design decides only one system message is supported, the extra ones are rejected with a 400 rather than dropped

## Tests to add

- [ ] `Tests/apfelTests/ContextManagerTests.swift` - `testAllSystemMessagesReachInstructions()` asserts both system contents appear in the built instructions
- [ ] `Tests/apfelTests/ContextManagerTests.swift` - `testNoSystemMessageIsUnchanged()` asserts the empty case is untouched
- [ ] `Tests/integration/server_validation_test.py` - `test_second_system_message_changes_prompt_tokens` asserts `prompt_tokens` grows when a second system message is added

---
<sub>Filed by automated review: OpenAI Codex `gpt-6-astra` (xhigh) reviewed, Claude Code verified against commit b39659abe6b627e95454c8144d8fdeff9b3f50a3 on 2026-09-06. Pass: edge-case.</sub>

### PR DESCRIPTION
**Pull request by Arthur-Ficial**
**Draft PR - needs @franzenzenhofer review and local test run before merging.**

Fixes #390.

## Summary

`ContextManager.buildInstructions` used `messages.first(where:)` to pick only the first system message. Since `makeSession` strips all system messages from the conversation array (line 35), every system message after the first was silently dropped - no error, no token accounting.

The fix adds `[OpenAIMessage].joinedSystemContent` in `ApfelCore` that filters, extracts, and joins all system-role text content with double newlines. `ContextManager` now calls this instead of `first(where:)`.

This affects all paths through `buildInstructions`: `/v1/chat/completions`, Responses `developer` messages (mapped to `system` by `ResponsesMapper`), and CLI `--messages` combined with `-s`.

## Root cause

`Sources/ContextManager.swift:128` called `messages.first(where: { $0.role == "system" })?.textContent` which returns only the first system message's text. Line 35 filters all system messages out of the conversation with `messages.filter { $0.role != "system" }`, so any system message not captured by `buildInstructions` vanishes completely - no token counting, no generation, no error.

## The fix

Added a public `joinedSystemContent` computed property on `[OpenAIMessage]` in `Sources/Core/OpenAIModels.swift` that collects all system messages, extracts their non-empty text content, and joins them with `\n\n`. `ContextManager.buildInstructions` now calls `messages.joinedSystemContent` instead of the single-message lookup. The logic lives in `ApfelCore` so it is unit-testable from the test runner.

## Test coverage

- Added 7 tests in `Tests/apfelTests/OpenAIModelsTests.swift` covering:
  - No system messages (nil)
  - Single system message (unchanged behavior)
  - Multiple system messages joined with `\n\n`
  - Empty-text system messages skipped
  - All-empty system messages (nil)
  - Nil-content system messages (nil)
  - Order preservation across interleaved roles

## What I verified (static only)

- Read `ContextManager.swift` end-to-end: the only consumer of system messages is `buildInstructions`, so the fix covers all paths.
- Read `ResponsesMapper.messages(from:)`: `developer` role maps to `system`, so Responses multi-developer conversations are fixed by the same change.
- Read `CLI.swift:messagesPrompt`: `-s` inserts a system message at index 0 before calling `makeSession`, so `--messages` + `-s` with an existing system message in the JSON is also fixed.
- Swift 6 concurrency: no data races introduced. The new extension is a pure computed property on a value type array.
- Diff limited to `Sources/ContextManager.swift`, `Sources/Core/OpenAIModels.swift`, `Tests/apfelTests/OpenAIModelsTests.swift`, `CHANGELOG.md` - no touches to release infrastructure, guardrails, CI, or dependencies.

## What I did NOT verify

- **Functional correctness. This needs `make test` on a Mac with Apple Intelligence.** I cannot run FoundationModels code. If the fix does not actually resolve the reported behaviour, that is on me to refine - ping me in a review comment.

## Anything that seemed suspicious

Nothing - straightforward bug. The issue was filed by the repo owner account.

## Test plan
- [ ] `swift build -c release`
- [ ] `swift run apfel-tests`
- [ ] `make install` and manual smoke test

---

cc @franzenzenhofer - draft stays draft until you review, test locally, and mark ready.

_Generated by the apfel bug-solver routine._

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01RSZyVodoaRkSGvFiuaUeVn

---
_Generated by [Claude Code](https://claude.ai/code/session_01RSZyVodoaRkSGvFiuaUeVn)_

### CODE REVIEW

**Review by Arthur-Ficial — COMMENTED** [review]
## First-pass triage review — automated daily pass, 2026-09-07

Reviewing against the golden goal (UNIX tool + OpenAI-compatible server + chat, 100% on-device) and the PR process in `CLAUDE.md`. This is a **first-pass safety net, not an approval** — I have not merged, approved, or pushed anything.

**Classification:** Code: network/parsing/auth · `+80/-2` across 4 files · targets #390

### Process checks (`CLAUDE.md` §7)
- ✅ Code change ships with tests.
- ✅ `CHANGELOG.md` touched (the `changelog-gate` requirement for any `Sources/**` change).

### CI
All checks green. Note this proves less than usual right now: ~⅓ of runs in this repo are currently failing on a flaky `pip3 install` step (see #438), so a green run partly means *this run got a clean runner*. It does not substitute for a local `swift run apfel-tests`.

### ⚠️ 2 competing candidates for #390

The auto-fix routine opened more than one PR for this issue. Only one should land; the rest should be closed.

| PR | size | tests | CI | 
| --- | --- | --- | --- |
| **#403 (this one)** | `+80/-2` | yes | 🟢 green |
| #400 | `+107/-2` | yes | 🔴 pip flake only |

Because the CI signal is unreliable right now, **do not use green vs red as the tiebreaker** between these. The #462 group is the cautionary precedent: all three candidates were plausible and CI-green, and all three were wrong (the guard turned a crash into silent schema truncation). Compare the diffs on their merits and run the tests locally.

**Status: draft, awaiting @franzenzenhofer.** Franz merges, Franz releases — always.

Cheers, Arthur

**Comment by Arthur-Ficial** [comment]
Closing: #390 is fixed on `main` in e359a2c.

**This PR's diff is the one that landed.** It was applied to a scratch tree, built, and verified against the issue's own reproduction before landing, then committed with the reasoning and the measurements in the commit message. Thanks -- this was the best of the candidates for this issue.

The landed change is verified, not just plausible: every fix in this sweep was built and re-run against the original reproduction before being committed.

_Closed as part of the 2026-09-07 backlog sweep: 29 issues fixed, 61 candidate PRs adjudicated._
